crates/shared/src/event_handling.rs

```diff
@@ -91,47 +96,57 @@ where
     }
 
     pub fn last_handled_block(&self) -> Option<BlockNumberHash> {
-        self.last_handled_block
+        self.last_handled_blocks.last().cloned()
     }
 
-    async fn event_block_range(&self) -> Result<RangeInclusive<BlockNumber>> {
-        // Instead of using only the most recent event block from the db we also store the last
-        // handled block in self so that during long times of no events we do not query needlessly
-        // large block ranges.
-        let (last_handled_block_number, last_handled_block_hash) = match self.last_handled_block {
-            Some(block) => block,
-            None => self.store.last_event_block().await?,
-        };
+    async fn event_block_range(
+        &self,
+    ) -> Result<(RangeInclusive<BlockNumber>, Vec<BlockNumberHash>)> {
         let current_block = self.block_retriever.current_block().await?;
-        let current_block_number = current_block
-            .number
-            .context("missing block number")?
-            .as_u64();
-        let from_block = if Some(current_block.parent_hash) == last_handled_block_hash {
-            // no reorg
-            (last_handled_block_number, last_handled_block_hash)
+        let handled_blocks = if self.last_handled_blocks.is_empty() {
+            vec![self.store.last_event_block().await?]
         } else {
-            (
-                last_handled_block_number.saturating_sub(MAX_REORG_BLOCK_COUNT),
-                None,
-            )
+            self.last_handled_blocks.clone()
         };
 
+        let current_block_number = current_block.number.context("missing number")?.as_u64();
+        let last_handled_block_hash = handled_blocks.last().unwrap().1.context("missing hash")?;
+
+        // handle special case which happens most of the time (no reorg, just one new block is added)
+        if current_block.parent_hash == last_handled_block_hash {
+            let current_block = (current_block_number, current_block.hash);
+            return Ok((
+                (BlockNumber::Latest(current_block)..=BlockNumber::Latest(current_block)),
+                vec![current_block],
```

Comment:
Shouldn't this be `[...handled_blocks, current_block]`? In other words, extend the handled blocks with the new one (to make detecting future re-orgs easier?).

Reply:
**Edit**: I see that "re-making" the `replacement_blocks` is done in the caller.